mlir/lib/CAPI/Dialect/Rock.cpp

```diff
@@ -0,0 +1,382 @@
+0,      0,      0,      0,      0,      0,      4.0}, {      0,      0,      0,
+0,      0,      0,      0,      0,      0,      0,      0,      0,      0, 0, 0,
+0,      1.0}, {      0,      0,      0,      0,      0,      0,      0,      0,
+0,      0,      0,      0,      0,      0,      0,      0,      1.0}};
+  // clang-format on
+
+  std::vector<float> variables = {
+      (float)in_channels, (float)in_h,        (float)in_w,
+      (float)fil_h,       (float)fil_w,       (float)out_channels,
+      (float)batchsize,   (float)group_count, (float)leftPadH,
+      (float)leftPadW,    (float)rightPadH,   (float)rightPadW,
+      (float)strideH,     (float)strideW,     (float)dilationH,
+      (float)dilationW};
+
+  // calculate prediction of each tuning parameter
+  std::vector<float> prediction;
+  uint32_t cLen = coeffs.size();
+  uint32_t vLen = variables.size();
+  for (int j = 0; j < cLen; j++) {
+    float pred = 0;
+    for (int i = 0; i < vLen; i++) {
+      pred += coeffs[j][i] * variables[i];
+    }
+    pred += coeffs[j][vLen]; // equals to coeffs[*].size - 1
+    prediction.push_back(pred);
+  }
+
+  // sort each parameter in distance order to the predition
+  for (int i = 0; i < prediction.size(); i++) {
+    std::sort(tParams[i].begin(), tParams[i].end(),
+              [&](uint32_t lhs, uint32_t rhs) {
+                float lDis = (float)lhs - prediction[i];
+                lDis *= lDis;
+                float rDis = (float)rhs - prediction[i];
+                rDis *= rDis;
+                return lDis < rDis;
+              });
+  }
+
+  // create space with sorted vectors
+  for (uint32_t t0 : tParams[0]) {
+    for (uint32_t t1 : tParams[1]) {
+      for (uint32_t t2 : tParams[2]) {
+        for (uint32_t t3 : tParams[3]) {
+          for (uint32_t t4 : tParams[4]) {
+            for (uint32_t t5 : tParams[5]) {
+              for (uint32_t t6 : tParams[6]) {
+                for (uint32_t t7 : tParams[7]) {
+                  std::vector<uint32_t> tParam = {t0, t1, t2, t3,
+                                                  t4, t5, t6, t7};
+                  newSpace->tuningRange.push_back(tParam);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  newSpace->numHeuristicQuick = 1;
+}
+*/
+
+struct rock::TunableParams *
+createTunableParams(rock::RockGemmWrapperInterface op) {
+  struct rock::TunableParams *newSpace;
+  newSpace = new rock::TunableParams();
+  newSpace->primaryOpType =  op.getKernelType();
+  // create range and heuristic
+  createGemmTuningRangeBF(newSpace, op);
+  return newSpace;
+}
+
+// ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+//////--------------- FIX! --------- Will be relocaetd  to
+/// rockTuningImpl.cpp----------///////////
+
+//////// This one to gemm wrapper interfasce
+// Stringfy given param
+std::string toPerfConfig(Attribute param) {
+  std::string result;
```

Comment:
Maybe a `SmallVectorImpl<char> &out`, and also `llvm::raw_svector_ostream`, so you can use `<<`

And probably you can go with `Twine`